### DISCUSSION TITLE
Poor WiFi performance or noise when using long wires

### DISCUSSION THREAD

**Discussion by Facetti** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#656] [Show and tell]
Using the orginal criss-cross wiring resulted in a bad picture quality, missing lines and artefacts.
Therefore I changed the wiring, that all is straight:
![ESP_64x64_cabling](https://github.com/user-attachments/assets/169fdedb-5025-4fa5-b52a-880c9d3fd640)
which resulted in a perfect picture.

Wiring:
![grafik](https://github.com/user-attachments/assets/f3bf085d-1a09-4a0b-82c1-f41aa6274fc8)

Code:
#define PANEL_WIDTH 64
#define PANEL_HEIGHT 64  	// Panel height of 64 will required PIN_E to be defined.
#define PANELS_NUMBER 2 	// Number of chained panels, if just a single panel, obviously set to 1

//--left row--
#define PIN_R1 32
#define PIN_B1 33
#define PIN_R2 25
#define PIN_B2 26
#define PIN_A 27
#define PIN_C 14
#define PIN_CLK 12
#define PIN_OE 13

//--right row--
#define PIN_G1 19
//GND -> connect to GND other side
#define PIN_G2 5
#define PIN_E 17
#define PIN_B 16
#define PIN_D 4
#define PIN_LAT 2
//GND -> connect to GND below

#define PANE_WIDTH PANEL_WIDTH * PANELS_NUMBER
#define PANE_HEIGHT PANEL_HEIGHT
.
.
.
 HUB75_I2S_CFG mxconfig;
  mxconfig.mx_width = PANEL_WIDTH;
  mxconfig.mx_height = PANEL_HEIGHT;
  mxconfig.chain_length = PANELS_NUMBER;
  
  mxconfig.gpio.r1 = PIN_R1;    //left row
  mxconfig.gpio.b1 = PIN_B1;
  mxconfig.gpio.r2 = PIN_R2;
  mxconfig.gpio.b2 = PIN_B2;
  mxconfig.gpio.a = PIN_A;
  mxconfig.gpio.c = PIN_C;
  mxconfig.gpio.clk = PIN_CLK;
  mxconfig.gpio.oe = PIN_OE;
  
  mxconfig.gpio.g1 = PIN_G1;    //right row
  // GND
  mxconfig.gpio.g2 = PIN_G2;    
  mxconfig.gpio.e = PIN_E;        
  mxconfig.gpio.b = PIN_B;
  mxconfig.gpio.d = PIN_D;
  mxconfig.gpio.lat = PIN_LAT;
  // GND

  // Display Setup
  dma_display = new MatrixPanel_I2S_DMA(mxconfig);

**Comment by mrcodetastic** [comment]
The wires are too long in any case. Max 10 cm.  20cm jumper cables are like a TV antenna and pick up interference and noise the LED panels don't like.

**Reply by Facetti** [reply]
at least in my case there are no artefacts, ghosting etc. 
I'm using two 64x64 displays daisy chained and the pictures are cristal clear.

BTW the suplied cable for connecting two pannels has over 15cm and also works well.

**Comment by ghmpi** [comment]
Funny I just saw this post.. I've come here with a similar issue... but while my display is mostly OK.. I've got a whole lot of WIFI interference! My wires are all less then 15mm long!

My ping rates are pretty bad most of the time, in the 100's of ms, with a smattering of packet loss as well. Bad enough at times that my sign, which displays weather data, has trouble contacting the weather service.  Here is the funny part, same ESP module, if I remove it and let it run by itself without the display, I have no problems!

My display is a standard everyday HUB75 module in P4 variety with the FM6124 ICs.  I have two chained.
My ESP32 is a Espressif FACTORY original devkit C module, wrover-e. I have several and they all do the same thing.
I have painstakingly reduced the wiring, by hand, to at most 15mm! I was hoping this would fix it, but no luck. I have tried adding a very good solid state capacitor of 470uf, and ceramic bypass caps, no luck. Grounded all the ground pins on the esp32, again same. I even used aluminum tape to shield and ground the entire bottom of the ESP module (you can see in the pictures, it is covered in black tape and grounded to the ground pin on the ESP), and even used some shielding around it with no luck. See attached pictures.

On the software side, I have tried different settings, such as clkphase, 8M 10M 15M and 20M i2s speeds, multiple latch_blanking values, and FM6126A setting as well (works, but same interference). Double buffering, single buffering, reducing color depth, and different brightness levels. All with similar results.

Came here to share my experience and maybe resolve my issue. I can open a new discussion if needed. 

Thank you for your time.

Regards,
-Moses

![IMG_20240823_211325269_HDR-small](https://github.com/user-attachments/assets/69016bc8-ad07-46f3-a1cb-87f365d2b775)
![IMG_20240823_211230515-small](https://github.com/user-attachments/assets/3dc1fa70-da1f-492b-b708-bb630cce1884)
![IMG_20240823_211157875_HDR-small](https://github.com/user-attachments/assets/93b8e469-79c3-436c-ac1e-c86785791496)
![IMG_20240823_211116337-small](https://github.com/user-attachments/assets/42240b1c-0f3f-431b-93ff-8feefa1e06f7)

**Reply by board707** [reply]
I would advise to use a two separate boards - one for work with RGB panel and a second to connect the net.

**Reply by mrcodetastic** [reply]
> Came here to share my experience and maybe resolve my issue. I can open a new discussion if needed.

Excellent work with being exhaustive with trying to find a possible root-cause.

What power supply are you using? Have you tried running it from a simple 5V USB battery pack?

It is possible noise can be introduced from a wall switched mode power supply as well, which will affect everything.

**Comment by ghmpi** [comment]
Ah, yes, the power supply. Almost forgot about that. I haven't focused on it, but thank you for reminding me, indeed something to look at. I have been using a switching Riden lab supply, but I have plenty of other, older/transformer/etc, based ones on the bench. I just tried it with one of the old transformer ones, and I can't say for certain it is making a difference.

Yesterday I installed an external antenna version of the same ESP32 module. I'm thinking this may get me a better signal to the ESP and possibly overwhelm the noise?

Of all the electronics issues I've always found RF and high speed switching interference problems to be difficult to troubleshoot, it doesn't help that not too many people have the expensive equipment to get a good view of things.

Is there a way to stop and restart the display output during runtime? This may help me troubleshoot. The plan is to get some PCBs printed soon, that may help things a bit more. I think we are dealing with a lot of variables here and every little bit helps. It's sometimes hard to quantify the effects, as I'm not sure if I'm dealing with local wifi interference or saturation as well. Time to find Esp modules in 5ghz versions.

If anyone is curious I make these into aviation weather displays with scrolling weather reports. Video here:
https://www.youtube.com/watch?v=f9lOoqnOjoM
I don't sell them, I just make them for friends. I have used various displays, such as 8x8 max7219 based dot displays and WS1812 displays. I've recently started using these HUB75 ones, and I like them so far, just have to get them running a little bit better.

I'll update this discussion with results, in case in helps someone else as well.

Thank you board707 and mrcodetastic for your comments! 

Regards,
-Moses

![IMG_20240824_103400469_HDR-small-antenna](https://github.com/user-attachments/assets/2e6f4390-acbc-42ff-a489-e37b4ccde08b)

**Reply by mrcodetastic** [reply]
That's super cool. If you're using a professional bench power supply then it is highly unlikely to be noise from the bench power supply.

I will look at the library and implement a restartDMAoutpu() function. Currently there's only stopDMAoutput().

Perhaps what you can do is 'fade to black' when an update is required, then do the WiFi stuff and then start outputting again. Perhaps not ideal, but better than having poor WiFi connectivity all the time.

Also, does it matter what is being shown on the screen. i.e. How many LEDs are being turned on and off? For example: If you have DMA output active... but it's only displaying a blank screen, is WiFi performance still crap? If so then it's definitely noise being propagated back from all the LEDS being stimulated.

FINALLY, have you considered simply buying one of these for a few bucks? These are an ESP32-S2 with a Logic Level converter between the ESP32-S2 and the HUB75 (3.3v to 5v) to drive the panels better and for isolation. Only downside, is you won't have any other 'spare' GPIOs. If you are simply driving a chain of panels only, I'd buy one of these and try that instead. No need to mess around with your own breadboard as well then. WF1 is only a few bucks.

https://github.com/mrcodetastic/HD-WF1-LED-MatrixPanel-DMA

**Comment by ghmpi** [comment]
Mrcodetastic,

I'll try stopDMAoutput() next time and see what happens. I have not tried experimenting with a blank display, but good idea, that may help.

I didn't know about those HD-WF1 board, they do look like a neat solution if they work right. I was thinking of wiring in logic converters as a long shot, I may just buy a couple of those boards to experiment with.

I put the quality bar higher when I make these for other people; as they usually leave my physical reach, and I prefer not to play tech support over the phone :)

Thanks again for your help!

Regards,
-Moses

**Comment by ghmpi** [comment]
Some updates for the curious. 

I tested with long wires, seeing what would happen if the esp32 was further away from the display, hard to tell the difference with wifi interference, but it did induce glitching in the display. I have not seen a good video out there on what happens with long wires, so here is my HUB75 glitching video :)
https://www.youtube.com/watch?v=jWXfdtsxo0o

Shielding the HUB75 display from the ESP32 seems to have made the biggest difference, see attached picture. I used aluminum tape and insulated it with more tape before installing it behind the connector where the ESP32 attaches. I soon realized in all this testing, that my biggest problem with the interference is not when the display is running and has already connected, but it's the INITIAL wifi connection that is more sensitive to interference. Once I shielded parts of the display, I noticed an immediate decrease in the time the initial wifi connection takes. 

A curious test was with the ESP module plugged into the HUB75, with no display libraries loaded, it seems I still had interference. It appeared the plain fact that the matrix and ESP were plugged into each other and powered was causing trouble. I tried setting all output pins that went to the matrix display to 0 (ground) (before loading of the libraries) to see if maybe floating inputs were the cause, but my results were inconclusive. I know from experience floating inputs can cause trouble, so this may be something to keep in mind for the future.

I tried the stopDMAoutput with the screen, I can't really tell if it made a difference, it's possible, but it's very slight.

I'm happy with the results from shielding the led matrix at this point. I may have more modules for a simultaneous A/B test in the near future and I may still try the HD-WF1 modules as well.

Thanks everyone!

Regards,
-Moses

![hub75-shielding-small](https://github.com/user-attachments/assets/d8e240eb-b3c4-448d-87dd-25f002890458)

**Reply by Facetti** [reply]
Great!
I'm using WiFi too and have had some connection issues when the ESP was to near to the display. 
I'll try your solution!

**Reply by mrcodetastic** [reply]
Good point about initiating WiFI **before** starting the DMA library. Like all wireless silicon, the ESP32 needs to perform some basic RF calibration activities when WiFi is enabled - these then set the scene for how it'll operate going forward.

Perhaps enabling the DMA output (and associated noise generated) screws up the RF calibration a bit, which then affects RF/WiFi performance on an ongoing basis.

> I tested with long wires, seeing what would happen if the esp32 was further away from the display, hard to tell the difference with wifi interference, but it did induce glitching in the display. I have not seen a good video out there on what happens with long wires, so here is my HUB75 glitching video :) https://www.youtube.com/watch?v=jWXfdtsxo0o

These panels are supposed to be driven with a 5V signal. It's by accident they work with 3.3v as well. The longer cable drops the voltage and signal quality outside tolerance for these panels so they then start displaying garbage.

**Reply by Facetti** [reply]
See this simple project for showing videos via WiFi on a ESP32 with two 64x64. There WiFi is also started before the DMA library.
https://github.com/Facetti/E1.31-sACN-ESP32-matrix-display 

Power supply: Meanwell MW LRS-50-5 (15€) 
![WhatsApp Bild 2024-08-27 um 18 51 42_47b7a44d](https://github.com/user-attachments/assets/7d903ab2-1597-4234-8ee6-df4e7442fcc4)

**Comment by rafma0** [comment]
**A software workaround, for setups you can't fix with wiring/shielding**

Same WiFi problem here with *no* wires: Adafruit MatrixPortal S3, which plugs straight onto the panel (antenna right next to the HUB75 connector) — nothing to shorten, yet the symptom is identical. (Not a HW/C person; figured this out with an AI assistant and tested it on hardware: ESP32-S3, 64×64.)

A/B test: DMA running → HTTPS fetches fail / ping loss (even at RSSI -40, `setSleep(false)`, min brightness). DMA never started → 100% OK. That's also why OTA works — `onStart` calls `stopDMAoutput()`.

Fix = "burst": stop the DMA only for the request, then resume: `stop DMA → fetch (radio quiet) → resume DMA → re-apply brightness → redraw`. 100% fetch success here vs ~0% with the DMA running.

And you can resume **without forking the lib**: `stopDMAoutput()` doesn't free the DMA descriptors (only the destructor's `release()` does), and `dma_bus` is `protected` while `dma_transfer_start()` is `public` — so a tiny subclass works (ESP32-S3):

```cpp
class Panel : public MatrixPanel_I2S_DMA {
public:
  Panel(const HUB75_I2S_CFG &cfg) : MatrixPanel_I2S_DMA(cfg) {}
  void resumeDMAoutput() { dma_bus.dma_transfer_start(); }
};

**Reply by mrcodetastic** [reply]
Well done. May you kindly raise a PR to integrate the resume DMA output call please? This could be useful for people who wish to reduce power draw of / 'turn off' HUB75 panels when not in use.

**Reply by rafma0** [reply]
Done, raised it as #939 👍

Adds a `resumeDMAoutput()` to restart DMA after `stopDMAoutput()`, no reboot needed. Works for turning the panel off when idle, and also for the WiFi workaround (stop, fetch, resume).

One correction to what I said earlier: turns out brightness is actually preserved across stop/resume (`resetbuffers()` re-applies it), so you only need to redraw your content after resuming, no need to re-set brightness.
